invenio_rdm_records/collections/models.py

```diff
@@ -0,0 +1,144 @@
+    def get(cls, id_):
+        """Get a collection tree by ID."""
+        return cls.query.get(id_)
+
+    @classmethod
+    def get_by_slug(cls, slug, community_id):
+        """Get a collection tree by slug."""
+        return cls.query.filter(
+            cls.slug == slug, cls.community_id == community_id
+        ).one_or_none()
+
+
+# Collection Table
+class Collection(db.Model, Timestamp):
+    """Collection model.
+
+    Indices:
+    - id
+    - collection_tree_id
+    - path
+    - slug
+    """
+
+    __tablename__ = "collections_collection"
+    __table_args__ = (
+        # Unique constraint on slug and tree_id. Slugs should be unique within a tree.
+        UniqueConstraint(
+            "slug", "tree_id", name="uq_collections_collection_slug_tree_id"
+        ),
+    )
+
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    slug = db.Column(db.String(255), nullable=False)
+    path = db.Column(db.Text, nullable=False, index=True)
+    tree_id = db.Column(
+        db.Integer, db.ForeignKey("collections_collection_tree.id"), nullable=False
+    )
+    title = db.Column(db.String(255), nullable=False)
+    search_query = db.Column("query", db.Text, nullable=False)
+    order = db.Column(db.Integer, nullable=True)
+    # TODO index depth
+    depth = db.Column(
+        db.Integer, db.Computed("array_length(string_to_array(path, ','), 1) - 2")
+    )  # TODO client side
+    num_records = db.Column(db.Integer, nullable=True)
+
+    # Relationship to CollectionTree
+    collection_tree = db.relationship("CollectionTree", back_populates="collections")
+
+    @classmethod
+    def create(cls, slug, path, title, search_query, ctree_or_id, **kwargs):
+        """Create a new collection."""
+        with db.session.begin_nested():
+            if isinstance(ctree_or_id, CollectionTree):
+                collection = cls(
+                    slug=slug,
+                    path=path,
+                    title=title,
+                    search_query=search_query,
+                    collection_tree=ctree_or_id,
+                    **kwargs,
+                )
+            elif isinstance(ctree_or_id, int):
+                collection = cls(
+                    slug=slug,
+                    path=path,
+                    title=title,
+                    search_query=search_query,
+                    tree_id=ctree_or_id,
+                    **kwargs,
+                )
+            else:
+                raise ValueError(
+                    "Either `collection_tree` or `collection_tree_id` must be provided."
+                )
+            db.session.add(collection)
+        return collection
+
+    @classmethod
+    def get(cls, id_):
+        """Get a collection by ID."""
+        return cls.query.get(id_)
+
+    @classmethod
+    def get_by_slug(cls, slug, tree_id):
+        """Get a collection by slug."""
+        return cls.query.filter(cls.slug == slug, cls.tree_id == tree_id).one_or_none()
```

Comment:
Do we return None, or should it fail when it's not found?

Reply:
I see the point of failing fast, however, I prefer to return `None` for separation of concerns. Whether the collection does not exist is an exception, I think it should be up to the business layer to decide. From a data point of view, it tried to fetch the collection but nothing was found. 

I agree that we need to throw an exception at some point to have the presentation layer(s) acting accordingly (and that's currently missing in this PR). I will add that part in the next iteration since I refactored part of the public API / service layers from Alex's comments.

Reply:
Maybe `None` works in this case, but you will see soon its limitations. With a None return value there are 2 main problems (and maybe more):
1. every time you call the func, you need to check if the return value is None or not. And the subsequent code will have to handle this case. So, as in your PR, you end up having `if` everywhere. Not very elegant.
2. even if less applicable in this func, in general you can't differentiate different type of unhappy path.

An exception is much more elegant because the caller will have many more options on how to handle the unhappy path.
I don't really understand what are concerns are better separate with returning None.

No problem with solving this later on, in subsequent PRs.